### ISSUE TITLE
Update not working correctly for NuGet.exe (2.8.6) with -Safe option

### ISSUE THREAD

**Issue by maartenba** [NuGet/Home#1651]
_From @fknx on October 28, 2015 14:6_

Hello,

I am using the NuGet.exe (2.8.6) to update my packages with the `-Safe` option. According to the [docs](https://docs.nuget.org/consume/command-line-reference#update-command) the `-Safe` option will not update a package to a higher major or minor version.

When I update a prerelease package with version `6.44.1-C80673` using the `-Safe` option I would expect it to update to a package with version `6.44.1`, `6.44.2` and so on (or any prerelease version with a later version than with `6.44.1-C80673` but starting with `6.44`). Unfortunately this is not the case. Instead, the packages are updated to the latest prerelease version available (in my case `6.45.0-C80620`).

I suspect this is a bug of the `-Safe` option in correlation with prerelease packages.

(Note: I am not using the `-Prerelease` option).

Best regards,
Frederik

_Copied from original issue: NuGet/NuGetGallery#2752_

**Comment by yishaigalatzer** [comment]
Is this using nuget.exe or the visual studio extension?

**Comment by fknx** [comment]
I am using the nuget.exe.

**Comment by fknx** [comment]
I've taken a look into the code and I suspect this has to do with the implementation of the `GetSafeRange` method from the `VersionUtility` class:

https://github.com/NuGet/NuGet2/blob/861741bd780225a542c73a1d0904dd33b33a7a3d/src/Core/Utility/VersionUtility.cs#L498-L506

It returns a `VersionSpec` object with a non-inclusive max version of the passed in major and minor version + 1, so in my case the non-inclusive max version would be `6.45.0`. As the prerelease version `6.45.0-C80620` is treated to be earlier than `6.45.0` it is included in the version range defined by the `VersionSpec` object. 

As the description of the `-Safe` option explicitly states that it won't upgrade to a version with a higher major or minor version I believe this behavior is incorrect.

**Comment by fknx** [comment]
In case the `GetSafeRange` method's implementation is the cause for this issue I would suggest the following change:

``` csharp
return new VersionSpec
{
    IsMinInclusive = true,
    MinVersion = version,
    MaxVersion = new SemanticVersion(version.Version.Major, version.Version.Minor + 1, 0, "a")
};
```

Using this modification the non-inclusive max version would be `6.45.0-a` using the previous example. As far as I can see there is no possible (valid) version starting with `6.45` which would be _lower_.

**Comment by yishaigalatzer** [comment]
Thanks for the detailed analysis. I wonder if this is fixed in 3.2. I'm on the road at the moment so you can find and try it at dist.nuget.org/index.html

If it still repros we should take a fix

Sent from my Windows Phone

**Comment by johnataylor** [comment]
This should not repro in 3.3. The code is actually very different. Here is the relevant code:

Firstly the -Safe flag is translated into VersionConstraints:

https://github.com/NuGet/NuGet.Client/blob/dev/src/NuGet.Clients/PackageManagement.PowerShellCmdlets/Cmdlets/UpdatePackageCommand.cs#L335

And secondly, every candidate Package we receive from the feed(s) is filtered according to these constraints:

https://github.com/NuGet/NuGet.Client/blob/dev/src/NuGet.Core/NuGet.PackageManagement/Resolution/PrunePackageTree.cs#L213

So this seems a direct expression in the code of the defined behavior of the flag. Though having said that, I agree that the fix for the old 2.n code looks correct to me, though certainly a change of behavior.

**Comment by yishaigalatzer** [comment]
@fknx pinging to see if you are still feeling this pain with NuGet 3.3?

**Comment by fknx** [comment]
@yishaigalatzer Sorry for responding this late. Unfortunately we have some issues with NuGet 3.3 preventing us from upgrading at this time, so in case there will be another release for the NuGet V2 I would appreciate it if this fix would be added to this release.

**Comment by yishaigalatzer** [comment]
No, we don't plan an update to 2.x nuget.exe, but we will gladly fix whatever is blocking you from moving over to 3.3 or 3.4, can you point to the issue?

**Comment by fknx** [comment]
We've had some issues with missing Windows 8.1 and Windows Phone 8.0/8.1 tools (the update crashed although our app projects did not use the package we were trying to update) but these seem to be gone by now. Issue #1798 is preventing us from using NuGet 3.3 right now, but as it looks like this will be fixed in 3.4.

Additionally, the `Prerelease` switch from the `update` [command](https://docs.nuget.org/consume/command-line-reference#update-command) seems to have a changed behavior. Previously you did not have to specify it when upgrading a prerelease package to another prerelease version, but now you need to specify it explicitly. In my opinion this is a really good change, but our build scripts must be adjusted accordingly.

**Comment by emgarten** [comment]
Thanks for reporting this issue. We have not been able to reproduce this issue. If you are still able to reproduce this with the latest NuGet version, please provide additional steps.
